frontend/controller/service-worker.js

```diff
@@ -30,8 +30,53 @@ window.addEventListener('beforeinstallprompt', e => {
   sbp('okTurtles.events/emit', PWA_INSTALLABLE)
 })
 
+const serviceWorkerMap = new WeakMap()
+const waitUntilSwReady = () => {
+  const promise = new Promise((resolve, reject) => {
+    const messageChannel = new MessageChannel()
+    messageChannel.port1.onmessage = (event) => {
+      if (event.data.type === 'ready') {
+        resolve()
+      } else {
+        reject(event.data.error)
+      }
+      messageChannel.port1.close()
+    }
+    messageChannel.port1.onmessageerror = () => {
+      reject(new Error('Message error'))
+      messageChannel.port1.close()
+    }
+    const oncontrollerchange = () => {
+      navigator.serviceWorker.removeEventListener('controllerchange', oncontrollerchange, false)
+      reject(new Error('Service worker controller changed'))
+      waitUntilSwReady().catch(e => {
```

Comment:
Could you add a comment above line 49 explaining what's going on here? E.g. why we're rejecting the promise and then calling `waitUntilSwReady` again? (i.e. for people like me who are unfamiliar with what `controllerchange` is and how this works)

Reply:
Simplified this a bit